### PR TITLE
Flamegraph: Add switch for color scheme by value or by package

### PR DESCRIPTION
**Pull request by aocenas**
Adds a button with dropdown to select how the items in flamegraph should be colored.

The current value representation is a bit minimalistic, by changing the gradient in the dot in the button but the color dot is what we use as a [color picker](https://developers.grafana.com/ui/canary/index.html?path=/story/pickers-and-editors-colorpicker--basic) so it is sort of consistent with other pickers.

https://github.com/grafana/grafana/assets/1014802/860edf20-75e2-4473-8d7c-e0cb2a9a1fa1

Note:
This is behind feature flag `flameGraphV2`

### CODE REVIEW

File: public/app/plugins/panel/flamegraph/flamegraphV2/components/FlameGraph/rendering.ts [hunk 1]
```diff
@@ -184,28 +204,36 @@ export function renderRect(
...
-  if (!rect.collapsed) {
-    ctx.stroke();
+  const color =
+    colorScheme === ColorScheme.ValueBased
+      ? getBarColorByValue(rect.ticks, totalTicks, rangeMin, rangeMax)
+      : getBarColorByPackage(rect.label, theme);
 
-    if (foundNames) {
-      ctx.fillStyle = foundNames.has(name) ? getBarColor(h, l) : colors[55];
+  if (foundNames) {
```
Comment (aocenas) [inline public/app/plugins/panel/flamegraph/flamegraphV2/components/FlameGraph/rendering.ts:212]: Just refactoring to separate decision on colors from how to render and simplify the Ifs a bit that way.

File: public/app/features/explore/TraceView/components/utils/color-generator.tsx [hunk 2]
```diff
@@ -132,19 +132,20 @@ export function getRgbColorByKey(key: string, theme: GrafanaTheme2): [number, nu
 }
 
 export function getFilteredColors(colorsHex: string[], theme: GrafanaTheme2) {
+  const filtered = [...colorsHex];
```
Comment (aocenas) [inline public/app/features/explore/TraceView/components/utils/color-generator.tsx:135]: @joey-grafana fyi did not end up using this for now but I thought it make sense to keep this fix here. As we just passed the colors from packages/grafana-ui/src/utils/colors.ts in the generator, this was modifying that global array which is used in other places which could make other colors glitch as other parts just picked colors by index.

File: public/app/plugins/panel/flamegraph/flamegraphV2/components/FlameGraph/murmur3.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+*/
+
+/* eslint-disable no-plusplus */
+/* eslint-disable prefer-const */
+/* eslint-disable no-bitwise */
+/* eslint-disable camelcase */
```
Comment (joey-grafana) [inline public/app/plugins/panel/flamegraph/flamegraphV2/components/FlameGraph/murmur3.ts:19]: Do we need all of these?

File: public/app/plugins/panel/flamegraph/flamegraphV2/components/FlameGraph/murmur3.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+export default function murmurhash3_32_gc(key: string, seed = 0) {
```
[the remark this answers is not in the document]
Comment (joey-grafana) [reply public/app/plugins/panel/flamegraph/flamegraphV2/components/FlameGraph/murmur3.ts:21]: If we only need a hash can we use the same one from the [color-generator](https://github.com/grafana/grafana/blob/main/public/app/features/explore/TraceView/components/utils/color-generator.tsx#L82)?

File: public/app/plugins/panel/flamegraph/flamegraphV2/components/FlameGraph/colors.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  }
+  return packageColor;
+}
+
+// const getColors = memoizeOne((theme) => getFilteredColors(colors, theme));
+
+// Different regexes to get the package name and function name from the label. We may at some point get an info about
+// the language from the backend and use the right regex but right now we just try all of them from most to least
+// specific.
+const matchers = [
```
Comment (joey-grafana) [inline public/app/plugins/panel/flamegraph/flamegraphV2/components/FlameGraph/colors.ts:69]: This is cool.

**Review by joey-grafana — COMMENTED** [review]
- Do we want to do something about the code scanning alerts?
- Clicking around focusing blocks seems to mess up the color scheme

https://github.com/grafana/grafana/assets/90795735/5ea5177a-ad8f-467c-94ef-8eccbdb104fe

File: public/app/plugins/panel/flamegraph/flamegraphV2/components/FlameGraph/murmur3.ts (hunk 3 above)
Comment (aocenas) [reply public/app/plugins/panel/flamegraph/flamegraphV2/components/FlameGraph/murmur3.ts:19]: yup we don't need those

**Comment by aocenas** [comment]
@joey-grafana 
> If we only need a hash can we use the same one from the [color-generator](https://github.com/grafana/grafana/blob/main/public/app/features/explore/TraceView/components/utils/color-generator.tsx#L82)?
Would like to leave it for later. The hash function in traceView seems very simplistic (not that I am expert on hashFunctions) so not sure if it does not make sense to use some normal hash function for this so the hashes (and colors) are more randomly distributed. Will take a look at that at some point though.

> Do we want to do something about the code scanning alerts?
If they do not block the PR would also leave it for now. The problem is I am not really sure how to improve that right away and the labels are mostly machine generated so it's not likely somebody imho that somebody would pass some very weird string here (and even so it will just be slow UI). Also this may possibly change a bit if we pass the language information from backend already.

> Clicking around focusing blocks seems to mess up the color scheme
Can you try again? Does not happen for me and I remember fixing issue like this in this PR (not sure when) so wonder if you may been on some older commit or something like that.

**Review by joey-grafana — APPROVED** [review]
Appears to have cleared itself up. All you seem to have done is pulled from main so not sure what was causing the issue in the video.

This looks brilliant btw 👍

**Pull request merged**